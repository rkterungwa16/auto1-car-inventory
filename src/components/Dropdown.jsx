import React from 'react'

class Dropdown extends React.Component {
  constructor (props) {
    super(props)
    this.state = {
      dropdownIsOpen: false,
      selectedOption: [this.props.defaultDropdownTitle ? this.props.defaultDropdownTitle : ''],
      content: []
    }
  }

  componentWillReceiveProps (nextProps) {
    if (nextProps.dropdownContent) {
      this.setState({
        content: nextProps.dropdownContent
      })
    }
  }

  toggleList () {
    this.setState(prevState => ({
      dropdownIsOpen: !prevState.dropdownIsOpen
    }))
  }

  selectOption (tab) {
    const {
      content
    } = this.state
    const state = content.map((value, index) => {
      const selectedState = {
        title: value,
        selected: tab === index ? !value.selected : false
      }
      return selectedState
    })

    const selected = state.filter(options => options.selected === true)
    this.setState({
      categories: state,
      selectedOption: selected
    })
    this.fireChangeEvent({ label: selected[0].title })
  }

  fireChangeEvent (currentSelected) {
    if (currentSelected !== this.state.selectedOption && this.props.onChange) {
      this.props.onChange(currentSelected)
    }
  }

  renderDropdownContents (contentValues) {
    const renderedList = contentValues.map((value, index) => (
      <div
        key={value}
        className={
          value.selected
            ? 'list__item selected'
            : 'list__item'
        }
        onClick={() => { this.selectOption(index) }}
      >{value}</div>
    ))
    return renderedList
  }

  render () {
    const {
      dropdownContent
    } = this.props

    const {
      dropdownIsOpen,
      selectedOption
    } = this.state
    return (
      <div className='dropdown'>
        <button
          className='dropdown__btn'
          onClick={this.toggleList.bind(this)}
        ><span>None</span><i className='fas fa-caret-down fa-2x' /></button>
        <span className='anp-dropdown--title'>
          { selectedOption[0].title }
        </span>
        <div
          className={
            dropdownContent
              ? 'dropdown__content dropdown__content--show'
              : 'dropdown__content'}
        >
          {
            dropdownIsOpen
              ? this.renderDropdownContents(dropdownContent)
              : null
          }
        </div>
      </div>
    )
  }
}

export default Dropdown
