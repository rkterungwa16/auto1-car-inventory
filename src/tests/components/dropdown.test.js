import React from 'react'
import Enzyme from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import Dropdown from '../../components/Dropdown'

jest.genMockFromModule('uniqid')
Enzyme.configure({ adapter: new Adapter() })
const { shallow } = Enzyme

const getFilterParams = jest.fn()
const wrapper = shallow(
  <Dropdown
    dropdownContent={['content']}
    defaultDropdownTitle={{ title: 'title' }}
    getFilterParams={getFilterParams}
  />
)
test('should render Dropdown correctly', () => {
  expect(wrapper).toMatchSnapshot()
})

test('should render Dropdown with alt data correctly', () => {
  wrapper.setProps({
    dropdownContent: ['another content'],
    defaultDropdownTitle: { title: 'another title' }
  })
  expect(wrapper).toMatchSnapshot()
})
test('should toggle open a dropdown menu', () => {
  const instance = wrapper.instance()
  expect(wrapper.state('dropdownIsOpen')).toBe(false)
  instance.toggleList()
  expect(wrapper.state('dropdownIsOpen')).toBe(true)
})

test('should format dropdown input data', () => {
  const instance = wrapper.instance()
  expect(instance.formatData(['value'])[0].title).toBe('value')
})

test('should select option', () => {
  wrapper.setProps({
    dropdownContent: ['title'],
    defaultDropdownTitle: { title: 'title' }
  })
  const instance = wrapper.instance()
  expect(wrapper.state('selectedOption')[0].title).toBe('title')
  instance.selectOption(0)
  expect(wrapper.state('selectedOption')[0].title).toBe('title')
})

test('should call filterParams method on selecting filter parameter', () => {
  wrapper.setProps({
    dropdownContent: ['car'],
    defaultDropdownTitle: { title: 'title' }
  })
  const instance = wrapper.instance()
  instance.fireChangeEvent('title')
  expect(getFilterParams).toHaveBeenCalled()
})

test('should render dropdown content', () => {
  const instance = wrapper.instance()
  const renderedDropdown = instance.renderDropdownContents([
    { id: 1, title: 'title', selected: false
    }])
  expect(typeof renderedDropdown).toBe('object')
})
