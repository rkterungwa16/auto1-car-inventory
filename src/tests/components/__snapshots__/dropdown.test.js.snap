// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Dropdown correctly 1`] = `
<div
  className="dropdown"
>
  <button
    className="dropdown__btn"
    onClick={[Function]}
  >
    <span
      className="dropdown__text--medium"
    />
    <i
      className="fas fa-caret-down fa-2x"
    />
  </button>
  <div
    className="dropdown__content"
  />
</div>
`;

exports[`should render Dropdown with alt data correctly 1`] = `
<div
  className="dropdown"
>
  <button
    className="dropdown__btn"
    onClick={[Function]}
  >
    <span
      className="dropdown__text--medium"
    />
    <i
      className="fas fa-caret-down fa-2x"
    />
  </button>
  <div
    className="dropdown__content"
  />
</div>
`;
